File: dynamic_programming/distinct_arrays/distinct_arrays.py
""" Solution Module """
class Solution:
    """ Solution Class """
    def run_solution(self, n):
        """ Solution Method """
        # n =2, [[1,1][2]]
        # n=3, [[1,1,1], [1,2], [2,1]]
        # n=5, [[1,1,1,1,1], [2,1,1,1], [1,2,1,1], [1,1,2,1], [1,1,1,2], 
        # [2,2,1], [2,1,2], [1,2,2]] 1,1,2,3,5


        # The solution is the count of fibanacci.
        # Recursive with memoization
        cache = [-1] * (n + 1)
        # def fib(n: int) -> int:
        #     if n <= 2:
        #         return n

        #     if cache[n] != -1:
        #         return cache[n]

        #     cache[n] = fib(n - 1) + fib(n - 2)
        #     return cache[n]
        # return fib(n)

        # Iterative
        def iterative(n:int) -> int:
            if n<=2:
                return n
            
            prev1 = 1
            prev2 = 2
            # Move the window one step
            for _ in range(3, n+1):
                curr = prev1 + prev2
                prev1 = prev2
                prev2 = curr

            return prev2
        return iterative(n)

        # Formula

File: dynamic_programming/distinct_arrays/distinct_arrays.ts
export class Solution {
    /**
    * @param {string} testInput
    * @return {any} Returns either boolean to validate the test cases or a custom result. 
    */
    run_solution(testInput: any) {
        
        // The solution is the number of fibanacci series
        // Recursion with Memoization
        const cache = [];
        // const fib = (n) => {
        //     if(n<=2) {
        //         return n
        //     }
        //     if(cache[n] !== undefined) {
        //         return cache[n]
        //     }
        //     cache[n] = fib(n-1) + fib(n-2)
        //     console.log(cache)
        //     return cache[n]
        // }
        // return fib(testInput)

        // Iterative 
        const iterative = (n) => {
            if(n<=2){
                return n
            }
            let p1 = 1;
            let p2 = 2;

            for(let i=3; i<n+1; i++){
                const curr = p1 +p2;
                p1 = p2;
                p2 = curr;
            }
            return p2
        }
        return iterative(testInput)

    }
}